### PR TITLE
Feature/2729 fetch translated exercises

### LINKED ISSUE #2729: Fetch translated exercises
Right now we only support officially internationalized KA sites like fr, pt-br and es. We wanna expand that to any user-translated exercises like de. We should also fetch translations for the Perseus exercises.

### PR DESCRIPTION
**Pull request by jesumer**
Hi @aronasorman,

This will fix #2729. The perseus strings used are downloaded from crowdin and these are approved strings.
I have already created a python script to wrap the non-translated perseus exercises.
Attached here is my screenshot of example string("Select the correct answer.").
screen shot 2015-02-05 at 2 48 20 pm
This is a example string of answer content which is from assessment item data from perseus exercise.
And my python script output:
screen shot 2015-02-05 at 2 49 37 pm

TODO:
- Loop over hints content on assessment item data.

### CODE REVIEW

File: kalite/topic_tools/__init__.py [hunk 1]
```diff
@@ -459,9 +459,28 @@ def get_assessment_item_data(request, assessment_item_id=None):
         return None
 
     # TODO (rtibbles): Enable internationalization for the assessment_items.
+    logging.info('Wrapping the Assessment Item with _()')
+    lang_code = i18n.get_default_language()
+    activate(lang_code)
```
Comment (cpauya) [inline kalite/topic_tools/__init__.py]: Do we need to activate the language here?  Isn't the language supposed to be active on the back-end already?

File: kalite/topic_tools/__init__.py [hunk 2]
```diff
@@ -459,9 +459,28 @@ def get_assessment_item_data(request, assessment_item_id=None):
...
+    item_data = json.loads(assessment_item['item_data'])
+    question_content = _(item_data['question']['content'])
+    answerarea_content = _(item_data['answerArea']['options']['content'])
+    # TODO (jesumer): Loop over hints.
+    hints_content = _(item_data['hints'][0]['content'])
```
Comment (cpauya) [inline kalite/topic_tools/__init__.py]: We need to loop over this array because the `hints` array may contain more than one element.

**Comment by cpauya** [comment]
@jesumer Can you put sample references here for the Perseus exercise that you tested?  Also include the Khan Academy's Crowdin approved translation for reference.

This will make testing this PR easy.

**Comment by jesumer** [comment]
Hi @aronasorman,

I believe this PR is fixed. Please review.

File: kalite/topic_tools/__init__.py [hunk 3]
```diff
@@ -460,8 +460,36 @@ def get_assessment_item_data(request, assessment_item_id=None):
 
     # TODO (rtibbles): Enable internationalization for the assessment_items.
 
+    try:
+        logging.info('Wrapping the Assessment Item with _()')
```
Comment (rtibbles) [inline kalite/topic_tools/__init__.py]: I think this logging level might be too high. Could we change this and the one below to logging.debug or get rid of the logging statements altogether?

**Comment by jesumer** [comment]
Hi @aronasorman,

This PR will require this [issue] (https://github.com/fle-internal/ka-lite-central/issues/225#issuecomment-74817316) in order to proceed and to test.

**Comment by arceduardvincent** [comment]
Hi @aronasorman and @rtibbles .

As I replicate this issue and review.
The image below demonstrated that the fetch translated exercise:

> 1. In the right pane you can see the translation that has been made e.g(Klicke auf die Grafik um die Bilder hinzuzufügen).

> 2.In the browser console you can see the highlighted line  reflecting the `item_data ` that translation was made and it reflect in our asssesment.json in its content .

Therefore I conclude that the fetch translated exercises work accordingly.  

screen shot 2015-02-18 at 2 23 24 pm


To Replicate:
1. At the central server, run `python manage.py update_language_packs --no-srts --no-dubbed -l de --no-kalite-trans` which will  generate the language pack.

2. At the distributed server, call `python manage.py languagepackdownload -l de --commandline`
which will download the language pack in the central server.

Note: The `master branch` of the `central-server` and the `staging server` can't replicate this process due to some issue regarding of the translation that is not included in the `.mo` files especially in the `perseus exercises`. I made a fix for this #226 . Use the branch that refereed in that fix .

File: kalite/topic_tools/__init__.py [hunk 4]
```diff
@@ -460,8 +460,36 @@ def get_assessment_item_data(request, assessment_item_id=None):
...
+            converted_content = _(content_value)
+            contents_dict['content'] = converted_content
+            new_hints.append(contents_dict)
+
+        item_data['hints'] = new_hints
+        item_data['question']['content'] = question_content
+        item_data['answerArea']['options']['content'] = answerarea_content
+        # dump data to make it to a proper json format.
+        assessment_item['item_data'] = json.dumps(item_data)
+        logging.info('Successfully wrapped New Assessment Item with _()')
```
Comment (aronasorman) [inline kalite/topic_tools/__init__.py]: I don't think we want this as part of our logs. At least make it `logging.debug`, or altogether remove it.

File: kalite/topic_tools/__init__.py [hunk 5]
```diff
@@ -460,8 +460,36 @@ def get_assessment_item_data(request, assessment_item_id=None):
...
+
+    except KeyError:
+        logging.log("Something wrong with the format on assessment items json.")
```
Comment (aronasorman) [inline kalite/topic_tools/__init__.py]: 1. This needs to be more informative: include the error message from the exception itself.
2. Fix the string: `"There is something wrong with the format of the assessment items: %s"`
3. Use `logging.error`. Or `logging.warning` if this is not a catastrophic error.

**Comment by aronasorman** [comment]
Also, please try to fix the merge conflict.

**Comment by jesumer** [comment]
Hi,

I have fixed already the changes you suggest. Thanks.

**Comment by aronasorman** [comment]
Seems like `ajax_test_messages` keeps failing here. Can you check why @jesumer? (cc @MCGallaspy)

**Comment by jesumer** [comment]
Yeah sure I will.

**Comment by jesumer** [comment]
Hi @aronasorman,

I found out the test and check it out. Here's my screenshot:
screen shot 2015-02-23 at 2 08 20 pm

I traced the test and I think the element here isn't exist:
screen shot 2015-02-23 at 2 11 11 pm

I don't have any idea about this maybe @MCGallaspy  know about it.

**Comment by MCGallaspy** [comment]
@jesumer @aronasorman one reason that test could be failing is if the content it's navigating to has been downloaded on the test server.

**Comment by MCGallaspy** [comment]
After some experimentation I think the issue is that the alert messages are not being displayed in this branch (but they are in develop), so something here must have broken the test. Hooray for regression tests! Will try to track down the breaking change.

File: kalite/main/api_resources.py [hunk 6]
```diff
@@ -299,9 +310,8 @@ def obj_get_list(self, bundle, **kwargs):
     def obj_get(self, bundle, **kwargs):
         content_id = kwargs.get("id", None)
         content = get_content_data(bundle.request, content_id)
-        # MUST: Include messages in the api call.
+
         if content:
-            content['messages'] = get_messages_for_api_calls(bundle.request)
```
Comment (MCGallaspy) [inline kalite/main/api_resources.py]: I think the removal of this line is why the given test is failing.

**Comment by MCGallaspy** [comment]
In fact, seems like this file has failed to incorporate changes since Feb 5... Compare the histories in [here](https://github.com/mrpau/ka-lite/commits/feature/2729-fetch-translated-exercises/kalite/main/api_resources.py) and [here](https://github.com/learningequality/ka-lite/blob/develop/kalite/main/api_resources.py). So maybe if you pull the latest changes to develop and rebase, it will resolve this issue? @aronasorman any advice on fixing it up?

File: data/khan/assessmentitems.json [hunk 7]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (rtibbles) [inline data/khan/assessmentitems.json]: This needs to be removed. Assessmentitems.json is in .gitignore on mainline develop and should not be included.

**Comment by rtibbles** [comment]
I agree with @MCGallaspy - this needs to be rebased, because as it stands, it is regressing a bunch of changes in mainline develop.

**Comment by jesumer** [comment]
All is well. I had rebased this to our new pulled develop branch.

**Comment by aronasorman** [comment]
I'm wary of so many commits and files changed. Can you just cherry-pick the actual commits that accomplish this PR to a new branch?

**Comment by aronasorman** [comment]
Closing this for now.

**Pull request closed without merging**